### PR TITLE
ROX-25722: Add metrics to sensor registry store

### PR DESCRIPTION
**Pull request by dcaravel**
### Description

This PR adds metrics for various aspects of Sensors registry store to assist troubleshooting and provide insight into the current state of the store.

PR stacked on top of:
- https://github.com/stackrox/stackrox/pull/12360

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

- [x] added unit tests
- [x] modified existing tests

#### How I validated my change

Deployed ACS to OCP 4.16 infra cluster, setup port-forward to sensor metrics port (9090), and observed metrics with pull secrets by name feature on / off:

With Pull Secrets by Name DISABLED:

```sh
$ curl 127.0.0.1:9090/metrics | grep registry_store

# HELP rox_sensor_registry_store_central_integrations_count Current number of stored image integrations from Central
# TYPE rox_sensor_registry_store_central_integrations_count gauge
rox_sensor_registry_store_central_integrations_count 8
# HELP rox_sensor_registry_store_cluster_local_hosts_count Current number of cluster local (i.e. OCP Internal Registry) hosts inserted into the registry store
# TYPE rox_sensor_registry_store_cluster_local_hosts_count gauge
rox_sensor_registry_store_cluster_local_hosts_count 3
# HELP rox_sensor_registry_store_global_secret_entries_count Current number of stored global registry entries
# TYPE rox_sensor_registry_store_global_secret_entries_count gauge
rox_sensor_registry_store_global_secret_entries_count 4
# HELP rox_sensor_registry_store_pull_secret_entries_count Current number of stored pull secret entries
# TYPE rox_sensor_registry_store_pull_secret_entries_count gauge
rox_sensor_registry_store_pull_secret_entries_count 216
# HELP rox_sensor_registry_store_pull_secret_entries_size Rough size in bytes of the currently stored pull secret entries
# TYPE rox_sensor_registry_store_pull_secret_entries_size gauge
rox_sensor_registry_store_pull_secret_entries_size 344968
# HELP rox_sensor_registry_store_tls_check_count The total number of TLS checks requested via the registry store
# TYPE rox_sensor_registry_store_tls_check_count counter
rox_sensor_registry_store_tls_check_count 4
# HELP rox_sensor_registry_store_tls_check_duration_seconds Time taken in seconds to perform a TLS check on a registry host (does not include TLS check cache hits)
# TYPE rox_sensor_registry_store_tls_check_duration_seconds histogram
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.005"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.01"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.025"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.05"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.1"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.25"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="1"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="2.5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="10"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="+Inf"} 1
rox_sensor_registry_store_tls_check_duration_seconds_sum 0.076623415
rox_sensor_registry_store_tls_check_duration_seconds_count 1
```
`344817` Bytes ~= `0.3` MiB

`rox_sensor_registry_store_tls_check_count` - `rox_sensor_registry_store_tls_check_duration_seconds_count` = # of cache hits = 3

With Pull Secrets by Name ENABLED:

```sh
$ oc set env deploy/sensor ROX_SENSOR_PULL_SECRETS_BY_NAME=true
$ curl 127.0.0.1:9090/metrics | grep registry_store

# HELP rox_sensor_registry_store_central_integrations_count Current number of stored image integrations from Central
# TYPE rox_sensor_registry_store_central_integrations_count gauge
rox_sensor_registry_store_central_integrations_count 8
# HELP rox_sensor_registry_store_cluster_local_hosts_count Current number of cluster local (i.e. OCP Internal Registry) hosts inserted into the registry store
# TYPE rox_sensor_registry_store_cluster_local_hosts_count gauge
rox_sensor_registry_store_cluster_local_hosts_count 3
# HELP rox_sensor_registry_store_global_secret_entries_count Current number of stored global registry entries
# TYPE rox_sensor_registry_store_global_secret_entries_count gauge
rox_sensor_registry_store_global_secret_entries_count 4
# HELP rox_sensor_registry_store_pull_secret_entries_count Current number of stored pull secret entries
# TYPE rox_sensor_registry_store_pull_secret_entries_count gauge
rox_sensor_registry_store_pull_secret_entries_count 1129
# HELP rox_sensor_registry_store_pull_secret_entries_size Rough size in bytes of the currently stored pull secret entries
# TYPE rox_sensor_registry_store_pull_secret_entries_size gauge
rox_sensor_registry_store_pull_secret_entries_size 1.885448e+06
# HELP rox_sensor_registry_store_tls_check_count The total number of TLS checks requested via the registry store
# TYPE rox_sensor_registry_store_tls_check_count counter
rox_sensor_registry_store_tls_check_count 5
# HELP rox_sensor_registry_store_tls_check_duration_seconds Time taken in seconds to perform a TLS check on a registry host (does not include TLS check cache hits)
# TYPE rox_sensor_registry_store_tls_check_duration_seconds histogram
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.005"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.01"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.025"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.05"} 0
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.1"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.25"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="0.5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="1"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="2.5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="5"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="10"} 1
rox_sensor_registry_store_tls_check_duration_seconds_bucket{le="+Inf"} 1
rox_sensor_registry_store_tls_check_duration_seconds_sum 0.080492124
rox_sensor_registry_store_tls_check_duration_seconds_count 1
```

`1.885068e+06` Bytes ~= `1.8` MiB

`rox_sensor_registry_store_tls_check_count` - `rox_sensor_registry_store_tls_check_duration_seconds_count` = # of cache hits = 4

### CODE REVIEW

File: sensor/common/registry/tlscheck_cache_test.go
Comment (RTann) [inline sensor/common/registry/tlscheck_cache_test.go:14]: still surprised there is no testing context

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go:139]: > Will this mess with the metrics?

Yes intentionally - the metrics reset in`metrics.ResetRegistryMetrics()` are meant to represent 'current state', since the cleanup method zeros out the contents of the registry store, the associated metrics should be zeroed out as well.

> when is this called? 

It's called on sensor shutdown and I believe when Sensor re-establishes a connection to Central, below is the reverse call flow

https://github.com/stackrox/stackrox/blob/a26fd772cb732968e2672b3209b9753509639b5d/sensor/kubernetes/listener/resources/store_provider.go#L84-L88

https://github.com/stackrox/stackrox/blob/e8a3a75590d7c33236fbf7242abb4470cf149578/sensor/kubernetes/listener/listener_impl.go#L85-L91

Sensor Shutdown:
https://github.com/stackrox/stackrox/blob/e8a3a75590d7c33236fbf7242abb4470cf149578/sensor/kubernetes/eventpipeline/pipeline_impl.go#L116

Sensor Reconnect:
https://github.com/stackrox/stackrox/blob/e8a3a75590d7c33236fbf7242abb4470cf149578/sensor/kubernetes/eventpipeline/pipeline_impl.go#L131

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go:281]: Since the global pull secret registries are stored separately wanted to have them counted separately.  With this we'll know how many 'entries' exist in the global pull secret - which may aid in troubleshooting, for example, why authentication to mirrors are not working. ([OCP docs state](https://docs.openshift.com/container-platform/4.16/post_installation_configuration/connected-to-disconnected.html#:~:text=Add%20mirror%20registry%20credential%20to%20the%20cluster%20global%20pull%2Dsecret%3A) that mirror registries should have auth added to the global pull secret)

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go:532]: Yes, created a test for that specifically:

https://github.com/stackrox/stackrox/blob/dfa126024aa9795b0a898ff2d70c6569084b9971/sensor/common/registry/registry_store_test.go#L911-L912

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go:556]: I assumed the prometheus library was efficient enough to where that wouldn't be necessary, but suppose it doesn't hurt to only update at the end (outside of a few extra lines of code)

File: sensor/common/registry/registry_store_test.go
Comment (dcaravel) [reply sensor/common/registry/registry_store_test.go:834]: I chose `fake-thingy-name` name because it's the same length as `fake-secret-name` (therefore in the size tests  `entrySize` remains consistent) 😄

File: sensor/common/registry/tlscheck_cache.go
Comment (dcaravel) [reply sensor/common/registry/tlscheck_cache.go:97]: Yup!

I originally had a metric specifically for TLS Cache Miss Count, but removed it realizing the `registry_store_tls_check_duration_seconds` histogram includes a `_count` that also represents cache misses.

```
rox_sensor_registry_store_tls_check_duration_seconds_count 1
```

Can can add those two together to get total count.  There is an edge case where a few checks may not be recorded when a cache entry is first created, i.e. when multiple routines reach here:

https://github.com/stackrox/stackrox/blob/186a06bfd0f8732e2c09c1c22dd39ab0ed68f07e/sensor/common/registry/tlscheck_cache.go#L104-L105

Each routine will miss the cache at first, the first routine will trigger a TLS check, the others will used the cached value. I wasn't seeing an easy to way detect that.  Perhaps a 'total requests' counter is valuable in that case.

Re-thinking this, removing the 'cache hit count' and replacing with a 'total request count' is probably better, we can then take total request count - miss count derived from the histogram to get the true cache hits. Convinced myself - will the change.

File: pkg/registries/set_impl.go
Comment (dcaravel) [reply pkg/registries/set_impl.go:183]: https://github.com/stackrox/stackrox/pull/12379/commits/2099682c40bd98cf62e5640ed3b44afc0f5db067

File: sensor/common/registry/tlscheck_cache.go
Comment (dcaravel) [reply sensor/common/registry/tlscheck_cache.go]: https://github.com/stackrox/stackrox/pull/12379/commits/ab8151f20d90f5aa167d073112047243fd13b58c

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go]: [b5de717](https://github.com/stackrox/stackrox/pull/12379/commits/b5de7179bd3a1b5195e95c29b661ba0f510f06d4)

File: sensor/common/registry/registry_store.go
Comment (dcaravel) [reply sensor/common/registry/registry_store.go:281]: Ah, because this one we always have the full count avail so we 'Set' it.  The other (pull secrets) we wouldn't know the full count unless we looped through every namespace+secret+host (which would be expensive) - so instead we keep a 'running total' by increment/decrementing when each pull secret is processed.

File: sensor/common/registry/tlscheck_cache.go
Comment (RTann) [inline sensor/common/registry/tlscheck_cache.go:92]: I think a previous commit counted the number of cache hits. That may still be useful so we can compare how many cache hits we get vs the total number of calls

**Review by RTann — COMMENTED** [review]
Just one question otherwise LGTM

Is the plan to merge this after the base PR is merged? I'd lean towards doing that to minimize reviewing complications

**Comment by dcaravel** [comment]
> Is the plan to merge this after the base PR is merged? I'd lean towards doing that to minimize reviewing complications

Yes, all merges will go to master.

Will merge the base PR first, then the base of this PR should auto-switch master, I'll address any conflicts, merge that to master, and so on.

File: sensor/common/registry/tlscheck_cache.go
Comment (dcaravel) [reply sensor/common/registry/tlscheck_cache.go:92]: This 'new' way does allow for calculating hits and is more accurate then in the previous PR ([explanation here](https://github.com/stackrox/stackrox/pull/12379#discussion_r1717353976)), I had also added an example of how to do this in the PR description, TL;DR:

Cache | How
--- | ---
Total Requests | `rox_sensor_registry_store_tls_check_count`
Miss | `rox_sensor_registry_store_tls_check_duration_seconds_count`
Hit |  `rox_sensor_registry_store_tls_check_count` - `rox_sensor_registry_store_tls_check_duration_seconds_count`

**Comment by dcaravel** [comment]
Rebase of base PR broke this, working on it...

**Comment by dcaravel** [comment]
Rebased to address unit test failure.

File: pkg/registries/set_impl.go
Comment (RTann) [inline pkg/registries/set_impl.go:185]: I think this is a `RWMutex`, so we should use `RLock` and `RUnlock` here

File: sensor/common/registry/registry_store.go
Comment (RTann) [inline sensor/common/registry/registry_store.go:565]: these aren't used outside the `if`, so we can just declare them inside

**Review by lvalerom — APPROVED** [review]
LGTM

File: pkg/registries/set_impl.go
Comment (dcaravel) [reply pkg/registries/set_impl.go:160]: https://github.com/stackrox/stackrox/pull/12379/commits/2ab0e409b9a32342f509ad4af8fe71cc938b2f76

**Pull request merged**
